### PR TITLE
Switch OPA and Ranger authorizers to resolveSelections

### LINKED ISSUE #5439: Remove synthetic principals hack in Resolver
#5119 is introducing external principals, and is also introducing a temporary "hack" in the Resolver: synthetic principals and principal roles.

Synthetic principal and roles are a temporary workaround for external principals, until external authorizers change their implementation of `PolarisAuthorizer.resolveAuthorizationInputs()` to avoid calling `PolarisResolutionManifest.resolveAll()`, and instead only resolve the securables that are actually needed for authorization, cf. `PolarisResolutionManifest.resolveSelections()`.

### PR DESCRIPTION
**Pull request by zhang-arvin**
Fixes #5439

## Summary

Switch external authorizers (OPA, Ranger) from `resolveAll()` to `resolveSelections()` with `REFERENCE_CATALOG`, `REQUESTED_PATHS`, and `REQUESTED_TOP_LEVEL_ENTITIES`.

This avoids resolving the caller principal and principal roles, which external authorizers do not need and which may not exist in the metastore for external principals.

## Background

PR #5119 introduces external principals and a temporary hack in the Resolver: synthetic principals and principal roles. This hack is needed because the current external authorizers call `resolveAll()`, which tries to resolve the caller principal from the metastore — but external principals do not exist in the metastore.

## Changes

- **`OpaPolarisAuthorizer.resolveAuthorizationInputs()`**: Changed from `resolveAll()` to `resolveSelections()` with `REFERENCE_CATALOG`, `REQUESTED_PATHS`, `REQUESTED_TOP_LEVEL_ENTITIES`.
- **`RangerPolarisAuthorizer.resolveAuthorizationInputs()`**: Same change.

## Why this works

External authorizers (OPA, Ranger) do not use resolved caller principal/principal roles in their authorization logic. They derive actor information from the `AuthorizationRequest.principal()` directly. By skipping `CALLER_PRINCIPAL` and `CALLER_PRINCIPAL_ROLES` resolution, we avoid the need for synthetic principal entities in the Resolver.

## Checklist

- [x] I have read the Polaris contribution guidelines
- [x] My code follows the code style of this project
- [x] No new dependencies are introduced
- [x] This change does not affect internal authorizer behavior (PolarisAuthorizerImpl still uses resolveAll())

### CODE REVIEW

File: extensions/auth/opa/src/main/java/org/apache/polaris/extension/auth/opa/OpaPolarisAuthorizer.java [hunk 1]
```diff
@@ -124,15 +125,23 @@ public OpaPolarisAuthorizer(
...
   public void resolveAuthorizationInputs(
       @NonNull AuthorizationState authzState, @NonNull AuthorizationRequest request) {
-    authzState.getResolutionManifest().resolveAll();
+    authzState
+        .getResolutionManifest()
+        .resolveSelections(
+            Set.of(
+                Resolvable.REFERENCE_CATALOG,
+                Resolvable.REQUESTED_PATHS,
+                Resolvable.REQUESTED_TOP_LEVEL_ENTITIES));
```
Comment (ayushtkn) [inline extensions/auth/opa/src/main/java/org/apache/polaris/extension/auth/opa/OpaPolarisAuthorizer.java:144]: This selection set is not equivalent to `resolveAll()` when the manifest has no reference catalog.
`ResolvePlan.fromSelections()` throws `IllegalArgumentException` if `REFERENCE_CATALOG` or `REQUESTED_PATHS` is selected and the catalog name is `null`. `resolveAll()` already skips those in that case.
`PolarisAdminService` uses `newResolutionManifest(null)` for root ops, `principal/principal-role` ops, and root grants. Those paths will break for OPA.
Always resolve `REQUESTED_TOP_LEVEL_ENTITIES`; add `REFERENCE_CATALOG` and `REQUESTED_PATHS` only when a reference catalog is present. 
Also update `OpaPolarisAuthorizerTest.resolveAuthorizationInputsResolvesAll()`, which still verifies `resolveAll()`.

File: extensions/auth/ranger/src/main/java/org/apache/polaris/extension/auth/ranger/RangerPolarisAuthorizer.java [hunk 2]
```diff
@@ -76,15 +77,23 @@ public void setRealmContext(RealmContext aRealmContext) {
...
       @NonNull AuthorizationState authzState, @NonNull AuthorizationRequest request) {
-    authzState.getResolutionManifest().resolveAll();
+    authzState
+        .getResolutionManifest()
+        .resolveSelections(
+            Set.of(
+                Resolvable.REFERENCE_CATALOG,
+                Resolvable.REQUESTED_PATHS,
+                Resolvable.REQUESTED_TOP_LEVEL_ENTITIES));
   }
```
Comment (ayushtkn) [inline extensions/auth/ranger/src/main/java/org/apache/polaris/extension/auth/ranger/RangerPolarisAuthorizer.java:97]: Same issue as `OpaPolarisAuthorizer`: this unconditional set will throw on admin calls with a `null` catalog. Please apply the same catalog-conditional selections here.

Would also be good to add Ranger unit tests for `resolveAuthorizationInputs()` covering both catalog-scoped and null-catalog manifests. `testAuthzRoot` only exercises `authorizeOrThrow`, so it would not catch this.

One caveat: a test that mocks `PolarisResolutionManifest` (like `OpaPolarisAuthorizerTest.resolveAuthorizationInputsResolvesAll()`) never reaches the real `ResolvePlan.fromSelections()`, so the `IllegalArgumentException` can't surface. To actually pin this, either capture the argument and assert the selection set excludes `REFERENCE_CATALOG`/`REQUESTED_PATHS` when there's no catalog, or use a real manifest constructed with a `null` catalog name.

**Review by ayushtkn — COMMENTED** [review]
Thanx @zhang-arvin for the changes.

On scope: the title and Fixes #5439 read as though the synthetic-principal hack is being removed from the Resolver, but this PR only changes the two authorizers — and the hack isn't in main yet, since #5119 is still open. To be fair to the PR, #5439's body defines the fix as exactly this ("until external authorizers change their implementation of resolveAuthorizationInputs() to avoid calling resolveAll()"), so the mismatch is really between that issue's title and its body.

I'd still retitle to something like "Switch OPA and Ranger authorizers to resolveSelections", so it's clear what this delivers, and let #5439 close when #5119 lands without the workaround.

**Review by adutra — CHANGES_REQUESTED** [review]
This isn't a fix for #5439.

**Comment by zhang-arvin** [comment]
Updated per review feedback:

- `PolarisResolutionManifest` now exposes a public `getCatalogName()` getter.
- OPA and Ranger authorizers branch their `resolveSelections` set on catalog presence: with a reference catalog they select `REFERENCE_CATALOG`, `REQUESTED_PATHS` and `REQUESTED_TOP_LEVEL_ENTITIES`; for null-catalog (admin) manifests they select only `REQUESTED_TOP_LEVEL_ENTITIES`, avoiding the `IllegalArgumentException` from `ResolvePlan.fromSelections`.
- `OpaPolarisAuthorizerTest.resolveAuthorizationInputsResolvesAll` replaced with selection-set asserts for both the catalog and null-catalog branches; symmetric tests added to `RangerPolarisAuthorizerTest`.
- PR title updated to 'Switch OPA and Ranger authorizers to resolveSelections'.

Let me know if anything else is needed. Thanks!

**Review by flyingImer — APPROVED** [review]
No blocking code concerns from my side. I agree with keeping #5439 open until the synthetic-principal workaround is actually removed, so this should use Related to rather than Fixes.
